File: apps/mac-client/Sources/Network/JitterBuffer.swift
// JitterBuffer.swift — Frame reassembly and jitter buffering
// GridStreamer macOS Client
//
// Mirrors the C++ JitterBuffer from gridstreamer-viewer. Reassembles fragmented
// video frames from individual UDP packets and buffers them to smooth out
// network jitter. Frames are released in frame_number order once all fragments
// are received. Incomplete frames older than max age are dropped.

import Foundation

/// A single reassembled video frame ready for decoding.
struct AssembledFrame: Sendable {
    /// The header from the first fragment (contains frame metadata).
    let header: VideoPacketHeader
    /// The concatenated payload data from all fragments, in order.
    let data: Data
    /// Local monotonic timestamp (in microseconds) when the first fragment arrived.
    let firstArrivalUs: UInt64
}

/// Thread-safe jitter buffer that reassembles fragmented video packets into
/// complete frames and releases them in order.
final class JitterBuffer: @unchecked Sendable {

    // MARK: - Frame Assembly State

    /// Tracks the assembly state of a single video frame.
    private struct FrameAssembly {
        var header: VideoPacketHeader
        var fragments: [Data?]         // Indexed by fragment_index
        var fragmentsReceived: Int = 0
        var fragmentTotal: Int
        var firstArrivalUs: UInt64
        var complete: Bool = false

        init(header: VideoPacketHeader, fragmentTotal: Int, firstArrivalUs: UInt64) {
            self.header = header
            self.fragmentTotal = fragmentTotal
            self.fragments = [Data?](repeating: nil, count: fragmentTotal)
            self.firstArrivalUs = firstArrivalUs
        }

        var isComplete: Bool {
            fragmentsReceived >= fragmentTotal && fragmentTotal > 0
        }

        /// Concatenate all fragments into a single contiguous Data buffer.
        func assemble() -> Data? {
            guard isComplete else { return nil }
            var result = Data()
            for frag in fragments {
                guard let frag else { return nil }
                result.append(frag)
            }
            return result
        }
    }

    // MARK: - Properties

    private var frames: [UInt16: FrameAssembly] = [:]  // frame_number -> assembly
    private var nextReleaseFrame: UInt16 = 0
    private var firstFrame = true
    private let lock = NSLock()

    /// Target jitter buffer depth in milliseconds. Default: 4ms for low-latency gaming.
    var targetDepthMs: UInt32 = 4

    /// Maximum age of an incomplete frame before it is dropped (ms).
    var maxFrameAgeMs: UInt32 = 150

    /// Statistics
    private(set) var framesDropped: UInt64 = 0
    private(set) var framesCompleted: UInt64 = 0
    private(set) var fragmentsReceived: UInt64 = 0

    // MARK: - Push

    /// Push a received video packet fragment into the buffer.
    /// - Parameters:
    ///   - header: The deserialized video packet header.
    ///   - payload: The raw payload data for this fragment.
    func pushPacket(header: VideoPacketHeader, payload: Data) {
        lock.lock()
        defer { lock.unlock() }

        let frameNum = header.frameNumber
        let fragIndex = Int(header.fragmentIndex)
        let fragTotal = Int(header.fragmentTotal)

        guard fragTotal > 0, fragIndex < fragTotal else { return }

        fragmentsReceived += 1

        let nowUs = currentTimeUs()

        if firstFrame {
            nextReleaseFrame = frameNum
            firstFrame = false
        }

        // Create or update the frame assembly
        if frames[frameNum] == nil {
            frames[frameNum] = FrameAssembly(
                header: header,
                fragmentTotal: fragTotal,
                firstArrivalUs: nowUs
            )
        }

        guard var assembly = frames[frameNum] else { return }

        // Store fragment if not already received
        if assembly.fragments[fragIndex] == nil {
            assembly.fragments[fragIndex] = payload
            assembly.fragmentsReceived += 1

            if assembly.isComplete {
                assembly.complete = true
                framesCompleted += 1
            }

            frames[frameNum] = assembly
        }

        // Expire old incomplete frames
        expireOldFrames(nowUs: nowUs)
    }

    // MARK: - Pop

    /// Pop the next complete frame in frame_number order.
    /// Returns `nil` if no complete frame is available at the next expected position.
    func popFrame() -> AssembledFrame? {
        lock.lock()
        defer { lock.unlock() }

        // Try to release frames in order starting from nextReleaseFrame
        // Skip over missing frames if they have been expired
        for _ in 0..<frames.count + 1 {
            guard let assembly = frames[nextReleaseFrame] else {
                // Frame not in buffer — if there are newer complete frames,
                // skip this gap (frame was likely dropped or never sent)
                if hasNewerCompleteFrame(than: nextReleaseFrame) {
                    nextReleaseFrame &+= 1
                    continue
                }
                return nil
            }

            if assembly.complete, let data = assembly.assemble() {
                let frame = AssembledFrame(
                    header: assembly.header,
                    data: data,
                    firstArrivalUs: assembly.firstArrivalUs
                )
                frames.removeValue(forKey: nextReleaseFrame)
                nextReleaseFrame &+= 1
                return frame
            } else {
                // Frame exists but is incomplete — wait for more fragments
                return nil
            }
        }

        return nil
    }

    // MARK: - Buffer Stats

    /// Returns the number of complete frames waiting in the buffer.
    var completeFrameCount: Int {
        lock.lock()
        defer { lock.unlock() }
        return frames.values.filter(\.complete).count
    }

    /// Returns the total number of frames (complete + incomplete) in the buffer.
    var totalFrameCount: Int {
        lock.lock()
        defer { lock.unlock() }
        return frames.count
    }

    /// Estimated buffer depth in milliseconds based on timestamp spread.
    var bufferDepthMs: UInt32 {
        lock.lock()
        defer { lock.unlock() }

        guard frames.count >= 2 else { return 0 }

        var minTs: UInt32 = .max
        var maxTs: UInt32 = 0
        for assembly in frames.values {
            let ts = assembly.header.timestampUs
            if ts < minTs { minTs = ts }
            if ts > maxTs { maxTs = ts }
        }

        // Timestamps are in microseconds; convert spread to milliseconds
        let spreadUs = maxTs >= minTs ? maxTs - minTs : 0
        return spreadUs / 1000
    }

    /// Reset the buffer, dropping all frames.
    func reset() {
        lock.lock()
        defer { lock.unlock() }
        frames.removeAll()
        firstFrame = true
        nextReleaseFrame = 0
    }

    // MARK: - Private Helpers

    /// Check if there is any complete frame with a number higher than `frameNum`.
    private func hasNewerCompleteFrame(than frameNum: UInt16) -> Bool {
        for (num, assembly) in frames {
            // Use wrapping comparison for sequence number wrap-around
            let diff = Int16(bitPattern: num &- frameNum)
            if diff > 0 && assembly.complete {
                return true
            }
        }
        return false
    }

    /// Expire frames that are older than maxFrameAgeMs.
    private func expireOldFrames(nowUs: UInt64) {
        let maxAgeUs = UInt64(maxFrameAgeMs) * 1000

        var toRemove: [UInt16] = []
        for (frameNum, assembly) in frames {
            if !assembly.complete && (nowUs - assembly.firstArrivalUs) > maxAgeUs {
                toRemove.append(frameNum)
            }
        }

        for frameNum in toRemove {
            frames.removeValue(forKey: frameNum)
            framesDropped += 1

            // If we dropped the frame we were waiting for, advance
            if frameNum == nextReleaseFrame {
                nextReleaseFrame &+= 1
            }
        }
    }

    /// Current monotonic time in microseconds.
    private func currentTimeUs() -> UInt64 {
        var info = mach_timebase_info_data_t()
        mach_timebase_info(&info)
        let nanos = mach_absolute_time() * UInt64(info.numer) / UInt64(info.denom)
        return nanos / 1000
    }
}

File: libs/nvremote-viewer/src/transport/jitter_buffer.cpp
///////////////////////////////////////////////////////////////////////////////
// jitter_buffer.cpp -- Frame reassembly and jitter buffering
//
// Reassembles fragmented video frames from UDP packets and holds them
// in a time-ordered buffer to smooth out network jitter before releasing
// them for decoding.
///////////////////////////////////////////////////////////////////////////////

#include "jitter_buffer.h"

#include <cs/common.h>

#include <algorithm>
#include <cstring>

namespace cs {

// ---------------------------------------------------------------------------
// Constructor / Destructor
// ---------------------------------------------------------------------------

JitterBuffer::JitterBuffer() = default;

JitterBuffer::~JitterBuffer() = default;

// ---------------------------------------------------------------------------
// pushPacket
// ---------------------------------------------------------------------------

void JitterBuffer::pushPacket(const VideoPacketHeader& header,
                               const uint8_t* payload, size_t len) {
    std::lock_guard<std::mutex> lock(mutex_);

    uint16_t frame_num = header.frame_number;
    uint8_t frag_idx = header.fragment_index;
    uint8_t frag_total = header.fragment_total;

    // Sanity checks
    if (frag_total == 0 || frag_idx >= frag_total) {
        CS_LOG(WARN, "JitterBuffer: invalid fragment %u/%u for frame %u",
               frag_idx, frag_total, frame_num);
        return;
    }

    // Initialize sequence tracking on first packet
    if (first_frame_) {
        next_release_frame_ = frame_num;
        first_frame_ = false;
    }

    // Check if this frame is too old (behind the release pointer by more than
    // half the sequence space). This handles 16-bit wraparound.
    int16_t delta = static_cast<int16_t>(frame_num - next_release_frame_);
    if (delta < -100) {
        // Very old frame, discard
        return;
    }

    // Find or create the frame assembly
    auto& assembly = frames_[frame_num];

    // First fragment for this frame: initialize
    if (assembly.fragment_total == 0) {
        assembly.header = header;
        assembly.fragment_total = frag_total;
        assembly.fragments.resize(frag_total);
        assembly.first_arrival_us = getTimestampUs();
    }

    // Avoid duplicate fragments
    if (frag_idx < assembly.fragments.size() && !assembly.fragments[frag_idx].empty()) {
        return;  // Already received this fragment
    }

    // Store the fragment
    if (frag_idx < assembly.fragments.size()) {
        assembly.fragments[frag_idx].assign(payload, payload + len);
        assembly.fragments_received++;
    }

    // Check if frame is now complete
    if (!assembly.complete && assembly.isComplete()) {
        assembly.complete = true;
    }

    // Expire old incomplete frames periodically
    expireOldFrames();
}

// ---------------------------------------------------------------------------
// popFrame
// ---------------------------------------------------------------------------

bool JitterBuffer::popFrame(std::vector<uint8_t>& frame_data, VideoPacketHeader& header) {
    std::lock_guard<std::mutex> lock(mutex_);

    // Look for the next frame in sequence
    auto it = frames_.find(next_release_frame_);
    if (it == frames_.end()) {
        // Frame not yet received. Check if we should skip ahead
        // (if we have a complete frame further ahead and the current one is very late).
        if (!frames_.empty()) {
            auto first_it = frames_.begin();
            int16_t gap = static_cast<int16_t>(first_it->first - next_release_frame_);
            if (gap > 0 && gap < 100) {
                // Check if the earliest frame we have is complete and old enough
                uint64_t now = getTimestampUs();
                uint64_t age_ms = (now - first_it->second.first_arrival_us) / 1000;
                if (first_it->second.complete && age_ms > target_depth_ms_) {
                    // Skip to this frame (dropping the missing ones)
                    uint16_t skipped = first_it->first - next_release_frame_;
                    frames_dropped_ += skipped;
                    next_release_frame_ = first_it->first;
                    it = first_it;
                }
            }
        }

        if (it == frames_.end()) {
            return false;
        }
    }

    // Check if the frame is complete
    if (!it->second.complete) {
        // Check if we should release it anyway (too old)
        uint64_t now = getTimestampUs();
        uint64_t age_ms = (now - it->second.first_arrival_us) / 1000;
        if (age_ms < max_frame_age_ms_) {
            return false;  // Wait for more fragments
        }

        // Frame is too old and still incomplete, drop it
        CS_LOG(DEBUG, "JitterBuffer: dropping incomplete frame %u (%u/%u fragments, age=%llu ms)",
               it->first, it->second.fragments_received, it->second.fragment_total,
               static_cast<unsigned long long>(age_ms));
        frames_dropped_++;
        frames_.erase(it);
        next_release_frame_++;
        return false;
    }

    // Check jitter buffer depth: don't release if we haven't buffered enough
    uint64_t now = getTimestampUs();
    uint64_t age_ms = (now - it->second.first_arrival_us) / 1000;
    // For very low latency (performance mode), release immediately when complete
    // For balanced/quality, hold for target_depth_ms_
    // Skip this delay if the buffer is getting full (>5 frames waiting)
    uint32_t complete_count = 0;
    for (const auto& f : frames_) {
        if (f.second.complete) complete_count++;
    }
    if (age_ms < target_depth_ms_ && complete_count < 3) {
        return false;  // Hold in buffer
    }

    // Assemble the frame
    if (!assembleFrame(it->second, frame_data)) {
        CS_LOG(WARN, "JitterBuffer: frame assembly failed for frame %u", it->first);
        frames_.erase(it);
        next_release_frame_++;
        return false;
    }

    header = it->second.header;
    frames_.erase(it);
    next_release_frame_++;

    return true;
}

// ---------------------------------------------------------------------------
// getBufferDepthMs
// ---------------------------------------------------------------------------

uint32_t JitterBuffer::getBufferDepthMs() const {
    std::lock_guard<std::mutex> lock(mutex_);

    if (frames_.empty()) return 0;

    uint64_t now = getTimestampUs();
    uint64_t oldest = frames_.begin()->second.first_arrival_us;
    return static_cast<uint32_t>((now - oldest) / 1000);
}

// ---------------------------------------------------------------------------
// setTargetDepthMs
// ---------------------------------------------------------------------------

void JitterBuffer::setTargetDepthMs(uint32_t ms) {
    std::lock_guard<std::mutex> lock(mutex_);
    target_depth_ms_ = ms;
    CS_LOG(DEBUG, "JitterBuffer: target depth set to %u ms", ms);
}

// ---------------------------------------------------------------------------
// getCompleteFrameCount
// ---------------------------------------------------------------------------

uint32_t JitterBuffer::getCompleteFrameCount() const {
    std::lock_guard<std::mutex> lock(mutex_);
    uint32_t count = 0;
    for (const auto& pair : frames_) {
        if (pair.second.complete) count++;
    }
    return count;
}

// ---------------------------------------------------------------------------
// getFramesDropped
// ---------------------------------------------------------------------------

uint64_t JitterBuffer::getFramesDropped() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return frames_dropped_;
}

// ---------------------------------------------------------------------------
// expireOldFrames
// ---------------------------------------------------------------------------

void JitterBuffer::expireOldFrames() {
    // Called under lock

    uint64_t now = getTimestampUs();
    uint64_t max_age_us = static_cast<uint64_t>(max_frame_age_ms_) * 1000;

    auto it = frames_.begin();
    while (it != frames_.end()) {
        if (!it->second.complete) {
            uint64_t age = now - it->second.first_arrival_us;
            if (age > max_age_us) {
                CS_LOG(DEBUG, "JitterBuffer: expiring frame %u (age=%llu us, %u/%u frags)",
                       it->first, static_cast<unsigned long long>(age),
                       it->second.fragments_received, it->second.fragment_total);
                frames_dropped_++;
                it = frames_.erase(it);
                continue;
            }
        }
        ++it;
    }

    // Also limit total buffer size to prevent unbounded growth
    while (frames_.size() > 100) {
        auto oldest = frames_.begin();
        CS_LOG(WARN, "JitterBuffer: buffer overflow, dropping frame %u", oldest->first);
        frames_dropped_++;
        frames_.erase(oldest);
    }
}

// ---------------------------------------------------------------------------
// assembleFrame
// ---------------------------------------------------------------------------

bool JitterBuffer::assembleFrame(const FrameAssembly& assembly,
                                  std::vector<uint8_t>& out) const {
    // Calculate total size
    size_t total_size = 0;
    for (const auto& frag : assembly.fragments) {
        if (frag.empty()) {
            // Should not happen if frame is complete, but be safe
            CS_LOG(WARN, "JitterBuffer: missing fragment in 'complete' frame");
            return false;
        }
        total_size += frag.size();
    }

    // Assemble in order
    out.clear();
    out.reserve(total_size);

    for (const auto& frag : assembly.fragments) {
        out.insert(out.end(), frag.begin(), frag.end());
    }

    return true;
}

} // namespace cs
